src/java.base/share/classes/java/nio/BufferCleaner.java

```diff
@@ -0,0 +1,268 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.nio;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Objects;
+import sun.nio.Cleaner;
+
+/**
+ * BufferCleaner supports PhantomReference-based management of native memory
+ * referred to by Direct-XXX-Buffers. Unreferenced DBBs may be garbage
+ * collected, deactivating the associated PRefs and making them available for
+ * cleanup here.
+ *
+ * There is a configured limit to the amount of memory that may be allocated
+ * by DBBs. When that limit is reached, the allocator may invoke the garbage
+ * collector directly to attempt to trigger cleaning here, hopefully
+ * permitting the allocation to complete. Only if that doesn't free sufficient
+ * memory does the allocation fail.  See java.nio.Bits::reserveMemory() for
+ * details.
+ *
+ * One of the requirements for that approach is having a way to determine that
+ * deactivated cleaners have been cleaned. java.lang.ref.Cleaner doesn't
+ * provide such a mechanism, and adding such a mechanism to that class to
+ * satisfy this unique requirement was deemed undesirable. Instead, this class
+ * uses the underlying primitives (PhantomReferences, ReferenceQueues) to
+ * provide the functionality needed for DBB management.
+ */
+class BufferCleaner {
+    private static final class PhantomCleaner
+        extends PhantomReference<Object>
+        implements Cleaner
+    {
+        private final Runnable action;
+        // Position in the CleanerList.
+        CleanerList.Node node;
+        int index;
+
+        public PhantomCleaner(Object obj, Runnable action) {
+            super(obj, queue);
+            this.action = action;
+        }
+
+        public void clean() {
+            if (cleanerList.remove(this)) {
+                // If being cleaned explicitly by application, rather than via
+                // reference processing by BufferCleaner, clear the referent so
+                // reference processing is disabled for this object.
+                clear();
+                try {
+                    action.run();
+                } catch (Throwable x) {
+                    // Long-standing behavior: when cleaning fails, VM exits.
+                    if (System.err != null) {
+                        new Error("nio Cleaner terminated abnormally", x).printStackTrace();
+                    }
+                    System.exit(1);
```

Comment:
This is the same behavior as jdk.internal.ref.Cleaner, for which this class is
substituting in the new regime for DBB management. PR 22165 (and earlier
versions of this PR) put this in the DBB's Deallocator::run method, but I
think it's both clearer here, and better to leave the Deallocator as it was in
mainline and be more consistent with the mainline code.